File: src_php/Site/Parkbuehne.php
<?php

declare(strict_types=1);

namespace MetalLE\Site;

use MetalLE\Event\Event;
use MetalLE\Event\Location;
use MetalLE\Site\Shop\Mawi;
use Traversable;

class Parkbuehne implements Site
{
    public function __construct(
        private Location $location = new Location('pb', 'Parkbühne im Clara-Zetkin-Park', 'https://www.parkbuehne-leipzig.com'),
    ) {
    }

    public function getIterator(): Traversable
    {
        $http = new Crawler();
        $html = $http->get('https://www.parkbuehne-leipzig.com/wordpress/veranstaltungen/');
        if (preg_match_all(
            '#<article\s.*<img\s.*src="(.*)".*<h3\s.*<a href="(.*)">(.*)</a></h3>.*</article>#isU',
            $html,
            $matches,
            PREG_SET_ORDER
        )) {
            foreach ($matches as $key => $match) {
                if (preg_match('#, (\d\d? [a-z]{2,3} \d\d\d\d)#iU', $match[0], $dateMatch)) {
                    $matches[$key][4] = new \DateTimeImmutable($dateMatch[1]);
                } else {
                    unset($matches[$key]);
                }
            }

            $mawi = new Mawi('Parkbühne Clara-Zetkin-Park');
            foreach ($mawi->fetchDates() as $date) {
                foreach ($matches as $match) {
                    if ($date->format('Ymd') === $match[4]->format('Ymd')) {
                        yield new Event(
                            html_entity_decode($match[3]),
                            $date,
                            $this->location,
                            $match[2],
                            $match[1],
                        );
                        break;
                    }
                }
            }
        }
    }

    public function getLocations(): iterable
    {
        yield $this->location;
    }
}

File: src/site/parkbuehne.rs
use std::borrow::Borrow;

use html_escape::decode_html_entities;
use regex::Regex;

use crate::{Event, Http, Location, Site};
use crate::site::Filter;
use crate::site::mawi::Mawi;
use crate::tools::date::parse_german_date;

pub(crate) struct Parkbuehne<'l> {
    location: Location<'l, 'l, 'l>,
    insecure_http: Http,
}

impl<'a> Parkbuehne<'_> {
    pub fn new(insecure_http: Http) -> Self {
        Self {
            location: Location {
                slug: "pb",
                name: "Parkbühne im Clara-Zetkin-Park",
                website: "https://www.parkbuehne-leipzig.com",
            },
            insecure_http,
        }
    }
}

impl<'a> Site for Parkbuehne<'_> {
    fn get_location(&self) -> &Location {
        self.location.borrow()
    }

    fn fetch_events(&self, http: &Http) -> Vec<Event> {
        let html = http
            .get("https://www.parkbuehne-leipzig.com/wordpress/veranstaltungen/")
            .unwrap();
        let reg = Regex::new("(?is)<article\\s.*?<img\\s.*?src=\"(?P<img>.*?)\".*?<h3\\s.*?<a href=\"(?P<url>.*?)\">(?P<name>.*?)</a></h3>.*?, (?P<date>\\d\\d? [a-z]{2,3} \\d\\d\\d\\d).*?</article>").unwrap();
        let mawi = Mawi::new("Parkbühne Clara-Zetkin-Park", self.insecure_http.borrow());

        let mut result = Vec::new();
        for capture in reg.captures_iter(html.as_str()) {
            let date = parse_german_date(capture.name("date").unwrap().as_str());
            let evt = Event::new(
                decode_html_entities(capture.name("name").unwrap().as_str()).to_string(),
                date.and_hms_opt(0, 0, 0).unwrap(),
                self.location.borrow(),
                capture.name("url").unwrap().as_str().to_string(),
                Some(capture.name("img").unwrap().as_str().to_string()),
            );
            if mawi.is_it_metal(evt.borrow()) {
                result.push(evt);
            }
        }

        result
    }
}
